### DISCUSSION TITLE
U Value Estimation Performance on Spark

### DISCUSSION THREAD

**Discussion by mshearer0** [moj-analytical-services/splink#1223] [General]
Has anyone else experienced any performance issues estimating u values using Spark? Using the same dataset on higher performance machines i'm finding it many times slower than DuckDB. 

I'm using scala-udf-similarity-0.1.0_spark3.3.jar from https://github.com/moj-analytical-services/splink_scalaudfs

**Comment by mshearer0** [comment]
Substituting Levenshstein or Exact Match comparisons results in the same slow performance.

**Reply by NickCrews** [reply]
I've never used spark, so I can't help, sorry!

**Comment by RossKen** [comment]
What sizes of dataset are you comparing? For smaller datasets DuckDB is the fastest linker. Our general rule of thumb is to run on DuckDB for datasets up to ~1 million records, after which it is worth considering spark.

I can't remember the exact numbers, but we did some head to head comparisons on runtime and even for 1million records Spark took 3-4 times longer than duckdb for a dedupe.

**Comment by RobinL** [comment]
Hi @mshearer0.  Thanks very much for the report.  We exchanged emails, but I wanted to ensure there was a reply here too. You've hit an unfortunate edge case.  I don't believe this is actually specifically a Spark issue, just a more general bug.

I've raised an issue [here](https://github.com/moj-analytical-services/splink/issues/1225)
